File: installer-tui/tests/disks.rs
//! lsblk parsing + partition-suffix tests.

use dots_installer::disks::{parent_disk, parse_lsblk, Disk};

const FIXTURE: &str = include_str!("fixtures/lsblk.json");

#[test]
fn keeps_only_writable_physical_disks() {
    let disks = parse_lsblk(FIXTURE).unwrap();
    let paths: Vec<&str> = disks.iter().map(|d| d.path.as_str()).collect();
    // nvme kept; usb stick kept (removable, string "1" rm field);
    // zram/rom/loop and the read-only vda excluded
    assert_eq!(paths, vec!["/dev/nvme0n1", "/dev/sda"]);
}

#[test]
fn parses_model_and_removable_flag() {
    let disks = parse_lsblk(FIXTURE).unwrap();
    assert_eq!(disks[0].model, "Samsung SSD 980");
    assert!(!disks[0].removable);
    assert!(
        disks[1].removable,
        "string \"1\" rm field parses as removable"
    );
}

#[test]
fn human_size_renders_gib() {
    let d = Disk {
        path: "/dev/nvme0n1".into(),
        size_bytes: 512_110_190_592,
        model: String::new(),
        removable: false,
    };
    assert_eq!(d.human_size(), "476.9 GiB");
}

#[test]
fn rejects_garbage_json() {
    assert!(parse_lsblk("not json").is_err());
}

#[test]
fn parent_disk_strips_partition_suffixes() {
    assert_eq!(parent_disk("/dev/sda1"), "/dev/sda");
    assert_eq!(parent_disk("/dev/nvme0n1p2"), "/dev/nvme0n1");
    assert_eq!(parent_disk("/dev/mmcblk0p1"), "/dev/mmcblk0");
    assert_eq!(parent_disk("/dev/vda"), "/dev/vda");
    assert_eq!(parent_disk("/dev/nvme0n1"), "/dev/nvme0n1");
}

File: nix/home/quickshell/qml/installer/disks.js
// Enumerate installable target disks by parsing `lsblk -J`.
//
// Ported from rust/installer-tui/src/disks.rs; parseLsblk/autodetectDisk are
// exercised against the same checked-in rust/installer-tui/tests/fixtures/
// lsblk.json in tests/qml/tst_installer.qml, so a change here that disagrees
// with the Rust parser fails the same way a change to disks.rs would.
.pragma library

/// One GiB in bytes.
const GIB = 1024 * 1024 * 1024;
/// ESP/boot partition size in the disko layout.
const ESP_GIB = 2;
/// Floor for the btrfs root: the desktop closure alone is ~12 GiB.
const ROOT_GIB = 20;

/// Minimum target disk size for the disko layout (ESP + swap + root), in GiB.
function requiredGib(swapGib) {
    return ESP_GIB + swapGib + ROOT_GIB;
}

/// lsblk emits native booleans (util-linux >= 2.37) or "0"/"1" strings.
function flag(v) {
    if (typeof v === "boolean")
        return v;
    if (typeof v === "string")
        return v === "1" || v === "true";
    if (typeof v === "number")
        return v === 1;
    return false;
}

function sizeOf(v) {
    if (typeof v === "number")
        return v;
    if (typeof v === "string") {
        const n = parseInt(v.trim(), 10);
        return Number.isNaN(n) ? 0 : n;
    }
    return 0;
}

/// Parse `lsblk -J -b -d -o NAME,PATH,SIZE,MODEL,RM,TYPE,RO` output. Keeps
/// writable physical disks only: excludes non-"disk" types (rom, loop),
/// read-only devices, and zram. Throws on malformed input, matching
/// disks.rs's `Result` return via `anyhow::anyhow!`.
function parseLsblk(json) {
    let root;
    try {
        root = JSON.parse(json);
    } catch (e) {
        throw new Error("lsblk output is not valid JSON");
    }
    const devices = root && Array.isArray(root.blockdevices) ? root.blockdevices : null;
    if (!devices)
        throw new Error("lsblk output missing 'blockdevices'");

    const disks = [];
    for (const dev of devices) {
        if (dev.type !== "disk")
            continue;
        const name = dev.name || "";
        if (name.startsWith("zram") || flag(dev.ro))
            continue;
        const sizeBytes = "size" in dev ? sizeOf(dev.size) : 0;
        if (sizeBytes === 0)
            continue;
        const path = typeof dev.path === "string" ? dev.path : `/dev/${name}`;
        const model = (typeof dev.model === "string" ? dev.model : "").trim();
        disks.push({
            path,
            sizeBytes,
            model,
            removable: flag(dev.rm),
        });
    }
    return disks;
}

/// Choose one sufficiently large target, preferring the sole fixed disk.
/// Throws when no disk qualifies or more than one candidate remains — the
/// caller (Network.qml) catches this to fall back to the manual DiskSelect
/// screen, exactly as `main.rs` falls back on an `Err` from this function.
function autodetectDisk(disks, swapGib) {
    const need = requiredGib(swapGib);
    const eligible = disks.filter(d => d.sizeBytes >= need * GIB);
    const fixed = eligible.filter(d => !d.removable);
    const candidates = fixed.length === 0 ? eligible : fixed;

    if (candidates.length === 0)
        throw new Error(`no installable disk has the required ${need} GiB capacity`);
    if (candidates.length !== 1)
        throw new Error(`disk autodetection is ambiguous: ${candidates.map(d => d.path).join(", ")}`);
    return candidates[0];
}

/// "476.9 GiB" style rendering.
function humanSize(disk) {
    const gib = disk.sizeBytes / GIB;
    return `${gib.toFixed(1)} GiB`;
}

/// Strip a partition suffix: /dev/sda1 -> /dev/sda, /dev/nvme0n1p2 -> /dev/nvme0n1.
/// Heuristic fallback for excluding the live medium — the QML port's
/// findmnt/lsblk PKNAME lookup prefers the authoritative PKNAME column and
/// only falls back to this when that lookup comes back empty, same as
/// disks.rs's `live_medium_disk`.
function parentDisk(path) {
    let stripped = path;
    while (stripped.length > 0 && /[0-9]/.test(stripped[stripped.length - 1])) {
        stripped = stripped.slice(0, -1);
    }
    if (stripped.length === path.length || stripped === "/dev/")
        return path;

    // Digit-named disks (nvme0n1, mmcblk0) use a 'p' separator before the
    // partition number; only a trailing pN may be stripped from them.
    if (stripped.endsWith("p")) {
        const pre = stripped.slice(0, -1);
        if (pre.length > 0 && /[0-9]/.test(pre[pre.length - 1]))
            return pre;
    }

    // A remaining inner digit (nvme0n[1]) means the "suffix" was part of the
    // disk name itself, not a partition number.
    const base = stripped.includes("/") ? stripped.slice(stripped.lastIndexOf("/") + 1) : stripped;
    if (/[0-9]/.test(base))
        return path;
    return stripped;
}
